builtin/providers/vcd/resource_vcd_edgegateway_vpn.go

```diff
@@ -0,0 +1,283 @@
+		LocalID:            d.Get("local_id").(string),
+		LocalSubnet:        localSubnets,
+		Mtu:                d.Get("mtu").(int),
+		PeerID:             d.Get("peer_id").(string),
+		PeerIPAddress:      d.Get("peer_ip_address").(string),
+		PeerSubnet:         peerSubnets,
+		SharedSecret:       d.Get("shared_secret").(string),
+		IsEnabled:          true,
+	}
+
+	tunnels := make([]*types.GatewayIpsecVpnTunnel, 1)
+	tunnels[0] = tunnel
+
+	ipsecVPNConfig := &types.EdgeGatewayServiceConfiguration{
+		Xmlns: "http://www.vmware.com/vcloud/v1.5",
+		GatewayIpsecVpnService: &types.GatewayIpsecVpnService{
+			IsEnabled: true,
+			Tunnel:    tunnels,
+		},
+	}
+
+	log.Printf("[INFO] ipsecVPNConfig: %#v", ipsecVPNConfig)
+
+	err = retryCall(vcdClient.MaxRetryTimeout, func() *resource.RetryError {
+		edgeGateway.Refresh()
+		task, err := edgeGateway.AddIpsecVPN(ipsecVPNConfig)
+		if err != nil {
+			log.Printf("[INFO] Error setting ipsecVPNConfig rules: %s", err)
+			return resource.RetryableError(
+				fmt.Errorf("Error setting ipsecVPNConfig rules: %#v", err))
+		}
+
+		return resource.RetryableError(task.WaitTaskCompletion())
+	})
+	if err != nil {
+		return fmt.Errorf("Error completing tasks: %#v", err)
+	}
+
+	d.SetId(d.Get("edge_gateway").(string))
+
+	return resourceVcdEdgeGatewayVpnRead(d, meta)
+}
+
+func resourceVcdEdgeGatewayVpnDelete(d *schema.ResourceData, meta interface{}) error {
+	vcdClient := meta.(*VCDClient)
+
+	log.Printf("[TRACE] CLIENT: %#v", vcdClient)
+
+	vcdClient.Mutex.Lock()
+	defer vcdClient.Mutex.Unlock()
+
+	edgeGateway, err := vcdClient.OrgVdc.FindEdgeGateway(d.Get("edge_gateway").(string))
+
+	ipsecVPNConfig := &types.EdgeGatewayServiceConfiguration{
+		Xmlns: "http://www.vmware.com/vcloud/v1.5",
+		GatewayIpsecVpnService: &types.GatewayIpsecVpnService{
+			IsEnabled: false,
+		},
+	}
+
+	log.Printf("[INFO] ipsecVPNConfig: %#v", ipsecVPNConfig)
+
+	err = retryCall(vcdClient.MaxRetryTimeout, func() *resource.RetryError {
+		edgeGateway.Refresh()
+		task, err := edgeGateway.AddIpsecVPN(ipsecVPNConfig)
+		if err != nil {
+			log.Printf("[INFO] Error setting ipsecVPNConfig rules: %s", err)
+			return resource.RetryableError(
+				fmt.Errorf("Error setting ipsecVPNConfig rules: %#v", err))
+		}
+
+		return resource.RetryableError(task.WaitTaskCompletion())
+	})
+	if err != nil {
+		return fmt.Errorf("Error completing tasks: %#v", err)
+	}
+
+	d.SetId(d.Get("edge_gateway").(string))
+
+	if err != nil {
+		return fmt.Errorf("Error finding edge gateway: %#v", err)
+	}
+
+	return nil
+}
+
+func resourceVcdEdgeGatewayVpnRead(d *schema.ResourceData, meta interface{}) error {
+	vcdClient := meta.(*VCDClient)
+
+	edgeGateway, err := vcdClient.OrgVdc.FindEdgeGateway(d.Get("edge_gateway").(string))
+	if err != nil {
+		return fmt.Errorf("Error finding edge gateway: %#v", err)
+	}
+
+	egsc := edgeGateway.EdgeGateway.Configuration.EdgeGatewayServiceConfiguration.GatewayIpsecVpnService
+	for _, t := range egsc.Tunnel {
```

Comment:
From the above code I assume the length of `egsc.Tunnel` should only be `1`. However, if it is ever not `1`, this will only set the attributes from the last set in the slice. If a user manually added an additional tunnel to the VPN configuration, this may yield false results or an unwanted state diff.